druid/src/widget/tabs.rs

```diff
@@ -0,0 +1,960 @@
+                    child.paint_raw(ctx, &data.inner, env);
+                })
+            }
+        } else if let Some(ref mut child) = Self::child(&mut self.children, data.selected) {
+            child.paint_raw(ctx, &data.inner, env);
+        }
+    }
+}
+
+// This only needs to exist to be able to give a reasonable type to the TabScope
+pub struct TabsScopePolicy<TP> {
+    tabs_from_data: TP,
+    selected: TabIndex,
+}
+
+impl<TP> TabsScopePolicy<TP> {
+    pub fn new(tabs_from_data: TP, selected: TabIndex) -> Self {
+        Self {
+            tabs_from_data,
+            selected,
+        }
+    }
+}
+
+impl<TP: TabsPolicy> ScopePolicy for TabsScopePolicy<TP> {
+    type In = TP::Input;
+    type State = TabsState<TP>;
+    type Transfer = LensScopeTransfer<tabs_state_derived_lenses::inner, Self::In, Self::State>;
+
+    fn create(self, inner: &Self::In) -> (Self::State, Self::Transfer) {
+        (
+            TabsState::new(inner.clone(), self.selected, self.tabs_from_data),
+            LensScopeTransfer::new(Self::State::inner),
+        )
+    }
+}
+
+#[derive(Data, Copy, Clone, Debug, PartialOrd, PartialEq)]
+pub enum TabsOrientation {
+    Standard,
+    Turns(u8), // These represent 90 degree rotations clockwise.
+}
+
+#[derive(Data, Copy, Clone, Debug, PartialOrd, PartialEq)]
+pub enum TabsTransition {
+    Instant,
+    Slide(Nanos),
+}
+
+impl Default for TabsTransition {
+    fn default() -> Self {
+        TabsTransition::Slide(Duration::from_millis(250).as_nanos() as Nanos)
+    }
+}
+
+impl TabsTransition {
+    fn tab_changed(self, old: TabIndex, new: TabIndex) -> Option<TabsTransitionState> {
+        match self {
+            TabsTransition::Instant => None,
+            TabsTransition::Slide(dur) => Some(TabsTransitionState::new(old, dur, old < new)),
+        }
+    }
+}
+
+impl TabsOrientation {
+    pub fn rotate_and_box<W: Widget<T> + 'static, T: Data>(
+        self,
+        widget: W,
+        axis: Axis,
+        cross: CrossAxisAlignment,
+    ) -> Box<dyn Widget<T>> {
+        let turns = match self {
+            Self::Standard => match (axis, cross) {
+                (Axis::Horizontal, _) => 0,
+                (Axis::Vertical, CrossAxisAlignment::Start) => 3,
+                (Axis::Vertical, _) => 1,
+            },
+            Self::Turns(turns) => turns,
+        };
+
+        if turns == 0 {
+            Box::new(widget)
+        } else {
+            Box::new(widget.rotate(turns))
+        }
+    }
+}
+
+pub struct InitialTab<T> {
+    name: String,
+    child: SingleUse<Box<dyn Widget<T>>>, // This is to avoid cloning provided tabs
+}
+
+impl<T: Data> InitialTab<T> {
+    pub fn new(name: impl Into<String>, child: impl Widget<T> + 'static) -> Self {
+        InitialTab {
+            name: name.into(),
+            child: SingleUse::new(Box::new(child)),
+        }
+    }
+}
+
+enum TabsContent<TP: TabsPolicy> {
+    Building {
```

Comment:
Can you explain why you went this route instead of having a separate `TabsBuilder` struct? (I also didn't understand the need for a `Build` associated type on `TabsPolicy`, as opposed to using a separate builder.)

Reply:
I originally had a builder, but my impression was that the 'integrated builder' style was heavily favoured in druid.

Reply:
I had a builder, knew that it would be complained about due to the stated preference for integrated builder methods - its been explicitly talked about as a design choice several times. 

So I used standard Rust features to avoid it and fit the stated design choices. 

And now that is a point of contention, but so far the comments don't seem to have any actual content, other than 'I didn't get this then I did'. As far as I'm aware it acheives its goals from an API pov, and doesn't corrupt the 'we don't do explicit builders for widgets' stated design choice. 

🤷‍♂️ I don't care either way other than it always feels a bit ... suboptimal ... to go back to a previous version of something. This bit of the api might go away with Crochet anyway.

Reply:
I'm not super opinionated here. I think it's nice to match the existing API patterns, but if something else made more sense then I'm happy to talk about that. I don't mind this as is. My previous comment should be interpreted as, "this is fine if you think it's best, but if you did it reluctantly because you wanted to match the existing API then I'm open to discussion an alternative design."

Reply:
I think its worth trying this one, if we find its not working and there isn't a better idea it can always be moved back. 
I expect people are more likely to copy examples to get started and the example here ends up looking reasonably straight forward.